### PR TITLE
Amp slidescroll experiment - Using touch end to load the next slide faster on IOS.

### PR DESCRIPTION
**Pull request by sriramkrish85**
**In this PR**

- detect scroll end when touchend is available and use timeout as a backup (this should fix IOS elasticity issue 90% of the times)

**Yet to come**

- Move autoplay to base template.
- Hide scrollbars where possible (and yet allow scrolling)

Part 8 for - #3465

### CODE REVIEW

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 1]
```diff
@@ -102,6 +111,37 @@ export class AmpSlideScroll extends BaseCarousel {
 
     this.slidesContainer_.addEventListener(
         'scroll', this.scrollHandler_.bind(this));
+
+    if(this.hasNativeSnapPoints_) {
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: space

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 2]
```diff
@@ -92,6 +95,12 @@ export class AmpSlideScroll extends BaseCarousel {
     /** @private {?number} */
     this.scrollTimeout_ = null;
 
+    /** @private {?number} */
+    this.touchEndTimeout_ = null;
+
+    /** @private {?boolean} */
+    this.hasTouchMoved_ = null;
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: Let's make non-nullable and set to false.

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 3]
```diff
@@ -102,6 +111,37 @@ export class AmpSlideScroll extends BaseCarousel {
...
+  }
+
+  touchEndHandler_() {
+    if (this.hasTouchMoved_) {
+      if (this.scrollTimeout_) {
+        timer.cancel(this.scrollTimeout_);
+      }
+      // Timer that detects scroll end and/or end of snap scroll.
+      this.touchEndTimeout_ = timer.delay(() => {
+        const currentScrollLeft = this.slidesContainer_./*REVIEW*/scrollLeft;
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: OK

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 4]
```diff
@@ -174,24 +214,30 @@ export class AmpSlideScroll extends BaseCarousel {
     if (this.scrollTimeout_) {
       timer.cancel(this.scrollTimeout_);
     }
+
     const currentScrollLeft = this.slidesContainer_./*OK*/scrollLeft;
+    console.log(this.element.id, 'scrolling at ' + currentScrollLeft);
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: remove

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 5]
```diff
@@ -174,24 +214,30 @@ export class AmpSlideScroll extends BaseCarousel {
...
-      } else {
-        this.customSnap_(currentScrollLeft);
-      }
-    }, timeout);
+    if (!this.touchEndTimeout_) {
+      const timeout =
+          this.hasNativeSnapPoints_ ? NATIVE_SNAP_TIMEOUT : CUSTOM_SNAP_TIMEOUT;
+      // Timer that detects scroll end and/or end of snap scroll.
+      this.scrollTimeout_ = timer.delay(() => {
+        console.log(this.element.id, 'scroll Ended at ' + currentScrollLeft);
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slidescroll.js]: remove

**Comment by dvoytenko** [comment]
This looks good. Are tests coming?

**Comment by sriramkrish85** [comment]
PTAL

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**